File: loader.cs
using Reko.Core;
using System;
using System.Collections.Generic;
using System.IO;

namespace Nucleus
{
    public class Symbol
    {
        public enum SymbolType
        {
            SYM_TYPE_UKN = 0x000,
            SYM_TYPE_FUNC = 0x001
        };

        public Symbol() { type = SymbolType.SYM_TYPE_UKN; name = null; addr = 0; }

        public SymbolType type;
        public string name;
        public ulong addr;
    };

    public enum SectionType
    {
        NONE = 0,
        CODE = 1,
        DATA = 2
    }

    public partial class Section
    {
        public Section() { binary = null; type = 0; vma = 0; size = 0; bytes = null; }

        public bool contains(ulong addr) { return (addr >= vma) && (addr - vma < size); }
        public bool is_import_table() { return name == ".plt"; }

        public Binary binary;
        public string name;
        public SectionType type;
        public ulong vma;
        public ulong size;
        public byte[] bytes;
    };

    public class Binary
    {
        public
          enum BinaryType
        {
            BIN_TYPE_AUTO = 0,
            BIN_TYPE_RAW = 1,
            BIN_TYPE_ELF = 2,
            BIN_TYPE_PE = 3
        };
        public enum BinaryArch
        {
            ARCH_NONE = 0,
            ARCH_AARCH64 = 1,
            ARCH_ARM = 2,
            ARCH_MIPS = 3,
            ARCH_PPC = 4,
            ARCH_X86 = 5
        };

        public Binary() { type = (0); arch = (0); bits = (0); entry = (0); }

        public string filename;
        public BinaryType type;
        public string type_str;
        public BinaryArch arch;
        public IProcessorArchitecture reko_arch;
        public string arch_str;
        public uint bits;
        public ulong entry;
        public List<Section> sections = new();
        public List<Symbol> symbols = new();

        public void create_reko_architecture()
        {
            switch (arch)
            {
            case BinaryArch.ARCH_X86:
                reko_arch = X86.create_architecture(this);
                break;
            case BinaryArch.ARCH_MIPS:
                reko_arch = Mips.create_architecture(this);
                break;
            default:
                Log.print_err("Reko support for {0} not implemented yet.", arch);
                Environment.Exit(1);
                return;
            }
        }
    }

    //int  load_binary   (string &fname, Binary *bin, Binary::BinaryType type);
    //void unload_binary (Binary *bin);


    partial class Nucleus
    {
        static string[][] binary_types_descr = {
  new[] {"auto", "Try to automatically determine binary format (default)"},
  new[] {"raw" , "Raw binary (memory dump, ROM, network capture, ...)"},
  new[] {"elf" , "Unix ELF"},
  new[] {"pe"  , "Windows PE"},
};

        public static (string str, Binary.BinaryArch arch, string description)[] binary_arch_descr = {
  ("auto"    , Binary.BinaryArch.ARCH_NONE, "Try to automatically determine architecture (default)"),
  ("aarch64" , Binary.BinaryArch.ARCH_AARCH64, "aarch64 (experimental)"),
  ("arm"     , Binary.BinaryArch.ARCH_ARM,     "arm (experimental)"),
  ("mips"    , Binary.BinaryArch.ARCH_MIPS,    "mips (experimental)"),
  ("ppc"     , Binary.BinaryArch.ARCH_PPC, "ppc: Specify ppc-32 or ppc-64 (default ppc-64, experimental)"),
  ("x86"     , Binary.BinaryArch.ARCH_X86, "x86: Specify x86-16, x86-32 or x86-64 (default x86-64)"),
        };

        #region BFD
        public class bfd
        {
            internal static object error_no_error;
            internal static object target_unknown_flavour;

            internal static void init()
            {
                throw new NotImplementedException();
            }

            internal static bfd openr(string p, object o)
            {
                throw new NotImplementedException();
            }

            internal static object get_error()
            {
                throw new NotImplementedException();
            }

            internal static object errmsg(object v)
            {
                throw new NotImplementedException();
            }

            internal static bool check_format(bfd bin, int bfd_object)
            {
                throw new NotImplementedException();
            }

            internal static void set_error(object error_no_error)
            {
                throw new NotImplementedException();
            }

            internal static object get_flavour(bfd bin)
            {
                throw new NotImplementedException();
            }

            internal static long get_symtab_upper_bound(bfd bfd_h)
            {
                throw new NotImplementedException();
            }

            internal static long get_dynamic_symtab_upper_bound(bfd bfd_h)
            {
                throw new NotImplementedException();
            }

            internal static long canonicalize_dynamic_symtab(bfd bfd_h, asymbol[] bfd_dynsym)
            {
                throw new NotImplementedException();
            }

            internal static ulong asymbol_value(asymbol asymbol)
            {
                throw new NotImplementedException();
            }
        }

        public class asymbol { }

        const int bfd_object = 0x4F;

        #endregion

        static bool bfd_inited = false;

        static bfd open_bfd(string fname)
        {

            bfd bin;

            if (!bfd_inited) {
                bfd.init();
                bfd_inited = true;
            }

            bin = bfd.openr(fname, null);
            if (bin == null) {
                Log.print_err("failed to open binary '{0}' ({1})", fname, bfd.errmsg(bfd.get_error()));
                return null;
            }

            if (!bfd.check_format(bin, bfd_object)) {
                Log.print_err("file '{0}' does not look like a binary object ({1}), maybe load as raw?", fname, bfd.errmsg(bfd.get_error()));
                return null;
            }

            /* Some versions of bfd_check_format pessimistically set a wrong_format
             * error before detecting the format, and then neglect to unset it once
             * the format has been detected. We unset it manually to prevent problems. */
            bfd.set_error(bfd.error_no_error);

            if (bfd.get_flavour(bin) == bfd.target_unknown_flavour) {
                Log.print_err("unrecognized format for binary '{0}' ({1})", fname, bfd.errmsg(bfd.get_error()));
                return null;
            }

            //Log.verbose(2, "binary '{0}' has format '{1}'", fname, bin.xvec.name);

            return bin;
        }


        int
        load_symbols_bfd(bfd bfd_h, Binary bin)
        {
            /*
            int ret;
            long n, nsyms, i;
            asymbol[] bfd_symtab;
            Symbol sym;

            bfd_symtab = null;

            n = bfd.get_symtab_upper_bound(bfd_h);
            if (n < 0)
            {
                Log.Log.print_err("failed to read symtab ({0})", bfd.errmsg(bfd.get_error()));
                goto fail;
            }
            else if (n != 0)
            {
                try
                {
                    bfd_symtab = new byte[n];
                }
                catch
                {
                    Log.Log.print_err("out of memory");
                    goto fail;
                }
                nsyms = bfd_canonicalize_symtab(bfd_h, bfd_symtab);
                if (nsyms < 0) {
                    Log.Log.print_err("failed to read symtab (%s)", bfd_errmsg(bfd_get_error()));
                    goto fail;
                }
                for (i = 0; i < nsyms; i++) {
                    if (bfd_symtab[i].flags & BSF_FUNCTION) {
                        sym = new Symbol();
                        bin.symbols.Add(sym);
                        sym.type |= Symbol.SymbolType.SYM_TYPE_FUNC;
                        sym.name = bfd_symtab[i].name;
                        sym.addr = bfd.asymbol_value(bfd_symtab[i]);
                    }
                }
            }

            ret = 0;
            goto cleanup;

            fail:
            ret = -1;

            cleanup:
            return ret;
            */
            return 0;
        }


        int
        load_dynsym_bfd(bfd bfd_h, Binary bin)
        {
            /*
            int ret;
            long n, nsyms, i;
            asymbol[] bfd_dynsym;
            Symbol sym;

            bfd_dynsym = null;

            n = bfd.get_dynamic_symtab_upper_bound(bfd_h);
            if (n < 0) {
                Log.Log.print_err("failed to read dynamic symtab (%s)", bfd_errmsg(bfd_get_error()));
                goto fail;
            } else if (n != 0) {
                bfd_dynsym = (asymbol**)malloc(n);
                if (!bfd_dynsym) {
                    Log.print_err("out of memory");
                    goto fail;
                }
                nsyms = bfd.canonicalize_dynamic_symtab(bfd_h, bfd_dynsym);
                if (nsyms < 0) {
                    Log.Log.print_err("failed to read dynamic symtab (%s)", bfd_errmsg(bfd_get_error()));
                    goto fail;
                }
                for (i = 0; i < nsyms; i++) {
                    if (bfd_dynsym[i].flags & BSF_FUNCTION) {
                        bin.symbols.push_back(Symbol());
                        sym = &bin.symbols.back();
                        sym.type |= Symbol::SYM_TYPE_FUNC;
                        sym.name = string(bfd_dynsym[i].name);
                        sym.addr = bfd_asymbol_value(bfd_dynsym[i]);
                    }
                }
            }

            ret = 0;
            goto cleanup;

            fail:
            ret = -1;

            cleanup:
            if (bfd_dynsym) free(bfd_dynsym);
            return ret;
            */
            return 0;
        }

        static int
        load_sections_bfd(bfd bfd_h, Binary bin)
        {
            /*
            int bfd_flags, sectype;
            uint64_t vma, size;
            const char* secname;
            asection* bfd_sec;
            Section* sec;

            for (bfd_sec = bfd_h.sections; bfd_sec; bfd_sec = bfd_sec.next) {
                bfd_flags = bfd_get_section_flags(bfd_h, bfd_sec);

                sectype = Section::SEC_TYPE_NONE;
                if (bfd_flags & SEC_CODE) {
                    sectype |= Section::SEC_TYPE_CODE;
                } else if (bfd_flags & SEC_DATA) {
                    sectype |= Section::SEC_TYPE_DATA;
                } else {
                    continue;
                }

                vma = bfd_section_vma(bfd_h, bfd_sec);
                size = bfd_section_size(bfd_h, bfd_sec);
                secname = bfd_section_name(bfd_h, bfd_sec);
                if (!secname) secname = "<unnamed>";

                sec = new Section();
                bin.sections.Add(sec);

                sec.binary = bin;
                sec.name = string(secname);
                sec.type = sectype;
                sec.vma = vma;
                sec.size = size;
                sec.bytes = (uint8_t*)malloc(size);
                if (!sec.bytes) {
                    Log.print_err("out of memory");
                    return -1;
                }

                if (!bfd_get_section_contents(bfd_h, bfd_sec, sec.bytes, 0, size)) {
                    Log.print_err("failed to read section '%s' (%s)", secname, bfd_errmsg(bfd_get_error()));
                    return -1;
                }
            }
            */
            return 0;
        }


        static int load_binary_bfd(string fname, Binary bin, Binary.BinaryType type)
        {
            /*
  int ret;
  bfd *bfd_h;
  const bfd_arch_info_type *bfd_info;

  bfd_h = NULL;

  bfd_h = open_bfd(fname);
  if(!bfd_h) {
    goto fail;
  }

  bin.filename = string(fname);
  bin.entry    = bfd_get_start_address(bfd_h);

  bin.type_str = string(bfd_h.xvec.name);
  switch(bfd_h.xvec.flavour) {
  case bfd_target_elf_flavour:
    bin.type = Binary::BIN_TYPE_ELF;
    break;
  case bfd_target_coff_flavour:
    bin.type = Binary::BIN_TYPE_PE;
    break;
  case bfd_target_unknown_flavour:
  default:
    Log.print_err("unsupported binary type (%s)", bfd_h.xvec.name);
    goto fail;
  }

  bfd_info = bfd_get_arch_info(bfd_h);
  bin.arch_str = string(bfd_info.printable_name);
  switch(bfd_info.mach) {
  case bfd_mach_i386_i386:
    bin.arch = Binary::ARCH_X86; 
    bin.bits = 32;
    break;
  case bfd_mach_x86_64:
    bin.arch = Binary::ARCH_X86;
    bin.bits = 64;
    break;
  default:
    Log.print_err("unsupported architecture (%s)", bfd_info.printable_name);
    goto fail;
  }

  /* Symbol handling is best-effort only (they may not even be present) */
            /*
            load_symbols_bfd(bfd_h, bin);
            load_dynsym_bfd(bfd_h, bin);

            if(load_sections_bfd(bfd_h, bin) < 0) goto fail;

            ret = 0;
            goto cleanup;

          fail:
            ret = -1;

          cleanup:
            if(bfd_h) bfd_close(bfd_h);

            return ret;
                      */
            return 0;
        }


        static int
        load_binary_raw(string fname, Binary bin, Binary.BinaryType type)
        {
            bin.filename = fname;
            bin.type = type;
            bin.type_str = "raw";

            if (options.binary.arch == Binary.BinaryArch.ARCH_NONE) {
                Log.print_err("cannot determine binary architecture, specify manually");
                return -1;
            }
            bin.arch = options.binary.arch;
            bin.bits = options.binary.bits;
            bin.arch_str = binary_arch_descr[(int)options.binary.arch].Item1;
            bin.entry = 0;

            if (bin.bits == 0) {
                switch (bin.arch) {
                case Binary.BinaryArch.ARCH_X86:
                    bin.bits = 64;
                    break;
                default:
                    break;
                }
            }

            var sec = new Section();
            bin.sections.Add(sec);

            sec.binary = bin;
            sec.name = "raw";
            sec.type = SectionType.CODE;
            sec.vma = options.binary.base_vma;

            try
            {
                //sec.bytes = generate_random_data(10);
                sec.bytes = File.ReadAllBytes(fname);
            }
            catch (Exception ex) {
                Log.print_err("failed to open binary '{0}' ({1})", fname, ex.Message);
                return -1;
            }

            sec.size = (uint)sec.bytes.Length;
            if (sec.size == 0)
            {
                Log.print_err("binary '{0}' appears to be empty", fname);
                return -1;
            }
            return 0;
        }

        private static byte[] generate_random_data(int size)
        {
            var mem = new byte[size];
            var rnd = new Random(42);
            rnd.NextBytes(mem);
            return mem;
            for (int i = 0; i < 10; ++i)
            {
                int c = rnd.Next() % 4;
                int p = rnd.Next() % mem.Length;
                for (int j = 0; j < c; ++j)
                {
                    mem[(p + j) % mem.Length] = 0x90;
                }
            }

            return mem;
        }

        static int load_binary(string fname, Binary bin, Binary.BinaryType type)
        {
            if (type == Binary.BinaryType.BIN_TYPE_RAW) {
                return load_binary_raw(fname, bin, type);
            } else {
                return load_binary_bfd(fname, bin, type);
            }
        }


        static void
        unload_binary(Binary bin)
        {
        }
    }
}

File: loader.cc
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <stdio.h>
#include <errno.h>

#include <string>
#include <vector>

#include <bfd.h>

#include "log.h"
#include "options.h"
#include "loader.h"


const char *binary_types_descr[][2] = {
  {"auto", "Try to automatically determine binary format (default)"},
  {"raw" , "Raw binary (memory dump, ROM, network capture, ...)"},
  {"elf" , "Unix ELF"},
  {"pe"  , "Windows PE"},
  {NULL  , NULL}
};

const char *binary_arch_descr[][2] = {
  {"auto"    , "Try to automatically determine architecture (default)"},
  {"aarch64" , "aarch64 (experimental)"},
  {"arm"     , "arm (experimental)"},
  {"mips"    , "mips (experimental)"},
  {"ppc"     , "ppc: Specify ppc-32 or ppc-64 (default ppc-64, experimental)"},
  {"x86"     , "x86: Specify x86-16, x86-32 or x86-64 (default x86-64)"},
  {NULL      , NULL}
};


static bfd*
open_bfd(std::string &fname)
{
  static int bfd_inited = 0;

  bfd *bin;

  if(!bfd_inited) {
    bfd_init();
    bfd_inited = 1;
  }

  bin = bfd_openr(fname.c_str(), NULL);
  if(!bin) {
    print_err("failed to open binary '%s' (%s)", fname.c_str(), bfd_errmsg(bfd_get_error()));
    return NULL;
  }

  if(!bfd_check_format(bin, bfd_object)) {
    print_err("file '%s' does not look like a binary object (%s), maybe load as raw?", fname.c_str(), bfd_errmsg(bfd_get_error()));
    return NULL;
  }

  /* Some versions of bfd_check_format pessimistically set a wrong_format
   * error before detecting the format, and then neglect to unset it once
   * the format has been detected. We unset it manually to prevent problems. */
  bfd_set_error(bfd_error_no_error);

  if(bfd_get_flavour(bin) == bfd_target_unknown_flavour) {
    print_err("unrecognized format for binary '%s' (%s)", fname.c_str(), bfd_errmsg(bfd_get_error()));
    return NULL;
  }

  verbose(2, "binary '%s' has format '%s'", fname.c_str(), bin->xvec->name);

  return bin;
}


int
load_symbols_bfd(bfd *bfd_h, Binary *bin)
{
  int ret;
  long n, nsyms, i;
  asymbol **bfd_symtab;
  Symbol *sym;

  bfd_symtab = NULL;

  n = bfd_get_symtab_upper_bound(bfd_h);
  if(n < 0) {
    print_err("failed to read symtab (%s)", bfd_errmsg(bfd_get_error()));
    goto fail;
  } else if(n) {
    bfd_symtab = (asymbol**)malloc(n);
    if(!bfd_symtab) {
      print_err("out of memory");
      goto fail;
    }
    nsyms = bfd_canonicalize_symtab(bfd_h, bfd_symtab);
    if(nsyms < 0) {
      print_err("failed to read symtab (%s)", bfd_errmsg(bfd_get_error()));
      goto fail;
    }
    for(i = 0; i < nsyms; i++) {
      if(bfd_symtab[i]->flags & BSF_FUNCTION) {
        bin->symbols.push_back(Symbol());
        sym = &bin->symbols.back();
        sym->type |= Symbol::SYM_TYPE_FUNC;
        sym->name = std::string(bfd_symtab[i]->name);
        sym->addr = bfd_asymbol_value(bfd_symtab[i]);
      }
    }
  }

  ret = 0;
  goto cleanup;

fail:
  ret = -1;

cleanup:
  if(bfd_symtab) free(bfd_symtab);

  return ret;
}


int
load_dynsym_bfd(bfd *bfd_h, Binary *bin)
{
  int ret;
  long n, nsyms, i;
  asymbol **bfd_dynsym;
  Symbol *sym;

  bfd_dynsym = NULL;

  n = bfd_get_dynamic_symtab_upper_bound(bfd_h);
  if(n < 0) {
    print_err("failed to read dynamic symtab (%s)", bfd_errmsg(bfd_get_error()));
    goto fail;
  } else if(n) {
    bfd_dynsym = (asymbol**)malloc(n);
    if(!bfd_dynsym) {
      print_err("out of memory");
      goto fail;
    }
    nsyms = bfd_canonicalize_dynamic_symtab(bfd_h, bfd_dynsym);
    if(nsyms < 0) {
      print_err("failed to read dynamic symtab (%s)", bfd_errmsg(bfd_get_error()));
      goto fail;
    }
    for(i = 0; i < nsyms; i++) {
      if(bfd_dynsym[i]->flags & BSF_FUNCTION) {
        bin->symbols.push_back(Symbol());
        sym = &bin->symbols.back();
        sym->type |= Symbol::SYM_TYPE_FUNC;
        sym->name = std::string(bfd_dynsym[i]->name);
        sym->addr = bfd_asymbol_value(bfd_dynsym[i]);
      }
    }
  }

  ret = 0;
  goto cleanup;

fail:
  ret = -1;

cleanup:
  if(bfd_dynsym) free(bfd_dynsym);

  return ret;
}


int
load_dynrelocs_bfd(bfd *bfd_h, Binary *bin)
{
  int ret;
  long nsyms, nrels, relsize, i;
  unsigned int symsize;
  asymbol **bfd_symtab = nullptr;
  arelent **bfd_relocs = nullptr;
  reloc_howto_type *bfd_howto;

  bfd_symtab = nullptr;
  bfd_relocs = nullptr;
  relsize = bfd_get_dynamic_reloc_upper_bound(bfd_h);
  if(relsize == 0) {
    return 0;
  }
  if(relsize < 0) {
    print_err("failed to read dynamic relocations size (%s)", bfd_errmsg(bfd_get_error()));
    goto fail;
  }
  nsyms = bfd_read_minisymbols(bfd_h, TRUE, (void**)&bfd_symtab, &symsize);
  if(nsyms < 0) {
    print_err("failed to read symtab (%s)", bfd_errmsg(bfd_get_error()));
    goto fail;
  }

  bfd_relocs = (arelent**)malloc(relsize);
  nrels = bfd_canonicalize_dynamic_reloc(bfd_h, bfd_relocs, bfd_symtab);
  if(nrels < 0) {
    print_err("failed to read dynamic relocations (%s)", bfd_errmsg(bfd_get_error()));
    goto fail;
  }
  /* Apply relocations */
  for(i = 0; i < nrels; i++) {
    arelent *bfd_reloc = bfd_relocs[i];
    asymbol *bfd_symbol = *(bfd_reloc->sym_ptr_ptr);
    bfd_howto = bfd_reloc->howto;

    for(const auto& sec : bin->sections) {
      /* Apply relocation to data of any executable section within range */
      size_t bytesize = (bfd_howto->bitsize / 8);
      if(bfd_reloc->address < sec.vma ||
         bfd_reloc->address > sec.vma + sec.size - bytesize ||
         sec.type != Section::SEC_TYPE_CODE) {
        continue;
      }
      /* Compute relocation value */
      bfd_vma relocation = 0;
      if(bfd_is_com_section(bfd_symbol->section)) {
        relocation = bfd_symbol->value;
      }
      relocation += bfd_reloc->addend;
      relocation >>= (bfd_vma)bfd_howto->rightshift;
      relocation <<= (bfd_vma)bfd_howto->bitpos;

      /* Patch data */
#define APPLY_RELOC(x) \
  ((x & ~bfd_howto->dst_mask) | (((x & bfd_howto->src_mask) + relocation) & bfd_howto->dst_mask))

      bfd_vma data_offset = bfd_reloc->address * bfd_octets_per_byte(bfd_h);
      bfd_byte* data = sec.bytes + (data_offset - sec.vma);

      switch (bfd_howto->size) {
      case 0: bfd_put_8 (bfd_h, APPLY_RELOC(bfd_get_8 (bfd_h, data)), data); break;
      case 1: bfd_put_16(bfd_h, APPLY_RELOC(bfd_get_16(bfd_h, data)), data); break;
      case 2: bfd_put_32(bfd_h, APPLY_RELOC(bfd_get_32(bfd_h, data)), data); break;
      case 4: bfd_put_64(bfd_h, APPLY_RELOC(bfd_get_64(bfd_h, data)), data); break;
      default:
        print_err("unsupported relocation size (%d)", bfd_howto->size);
        goto fail;
      }
#undef APPLY_RELOC
    }
  }

  ret = 0;
  goto cleanup;

fail:
  ret = -1;

cleanup:
  if(bfd_symtab) free(bfd_symtab);
  if(bfd_relocs) free(bfd_relocs);

  return ret;
}


int
load_sections_bfd(bfd *bfd_h, Binary *bin)
{
  int bfd_flags, sectype;
  uint64_t vma, size;
  const char *secname;
  asection* bfd_sec;
  Section *sec;

  for(bfd_sec = bfd_h->sections; bfd_sec; bfd_sec = bfd_sec->next) {
    bfd_flags = bfd_get_section_flags(bfd_h, bfd_sec);

    sectype = Section::SEC_TYPE_NONE;
    if(bfd_flags & SEC_CODE) {
      sectype |= Section::SEC_TYPE_CODE;
    } else if(bfd_flags & SEC_DATA) {
      sectype |= Section::SEC_TYPE_DATA;
    } else {
      continue;
    }

    vma     = bfd_section_vma(bfd_h, bfd_sec);
    size    = bfd_section_size(bfd_h, bfd_sec);
    secname = bfd_section_name(bfd_h, bfd_sec);
    if(!secname) secname = "<unnamed>";

    bin->sections.push_back(Section());
    sec = &bin->sections.back();

    sec->binary = bin;
    sec->name   = std::string(secname);
    sec->type   = sectype;
    sec->vma    = vma;
    sec->size   = size;
    sec->bytes  = (uint8_t*)malloc(size);
    if(!sec->bytes) {
      print_err("out of memory");
      return -1;
    }

    if(!bfd_get_section_contents(bfd_h, bfd_sec, sec->bytes, 0, size)) {
      print_err("failed to read section '%s' (%s)", secname, bfd_errmsg(bfd_get_error()));
      return -1;
    }
  }

  return 0;
}


int
load_binary_bfd(std::string &fname, Binary *bin, Binary::BinaryType type)
{
  int ret;
  bfd *bfd_h;
  const bfd_arch_info_type *bfd_info;

  bfd_h = NULL;

  bfd_h = open_bfd(fname);
  if(!bfd_h) {
    goto fail;
  }

  bin->filename = std::string(fname);
  bin->entry    = bfd_get_start_address(bfd_h);

  bin->type_str = std::string(bfd_h->xvec->name);
  switch(bfd_h->xvec->flavour) {
  case bfd_target_elf_flavour:
    bin->type = Binary::BIN_TYPE_ELF;
    break;
  case bfd_target_coff_flavour:
    bin->type = Binary::BIN_TYPE_PE;
    break;
  case bfd_target_unknown_flavour:
  default:
    print_err("unsupported binary type (%s)", bfd_h->xvec->name);
    goto fail;
  }

  bfd_info = bfd_get_arch_info(bfd_h);
  bin->arch_str = std::string(bfd_info->printable_name);
  switch(bfd_info->arch) {
  case bfd_arch_i386:
    switch(bfd_info->mach) {
    case bfd_mach_i386_i386:
      bin->arch = Binary::ARCH_X86;
      bin->bits = 32;
      break;
    case bfd_mach_x86_64:
      bin->arch = Binary::ARCH_X86;
      bin->bits = 64;
      break;
    default:
      goto fail_arch;
    }
    break;

  case bfd_arch_arm:
    switch(bfd_info->mach) {
    case bfd_mach_arm_5T:
      bin->arch = Binary::ARCH_ARM;
      bin->bits = 32;
      break;
    default:
      goto fail_arch;
    }
    break;

  case bfd_arch_aarch64:
    switch(bfd_info->mach) {
    case bfd_mach_aarch64:
    case bfd_mach_aarch64_ilp32:
      bin->arch = Binary::ARCH_AARCH64;
      bin->bits = 64;
      break;
    default:
      goto fail_arch;
    }
    break;

  case bfd_arch_mips:
    switch(bfd_info->mach) {
    case bfd_mach_mips16:
      bin->arch = Binary::ARCH_MIPS;
      bin->bits = 16;
      break;
    case bfd_mach_mipsisa32r2:
      bin->arch = Binary::ARCH_MIPS;
      bin->bits = 32;
      break;
    case bfd_mach_mipsisa64:
      bin->arch = Binary::ARCH_MIPS;
      bin->bits = 64;
      break;
    default:
      goto fail_arch;
    }
    break;

  case bfd_arch_powerpc:
    switch(bfd_info->mach) {
    case bfd_mach_ppc:
      bin->arch = Binary::ARCH_PPC;
      bin->bits = 32;
      break;
    case bfd_mach_ppc64:
      bin->arch = Binary::ARCH_PPC;
      bin->bits = 64;
      break;
    default:
      goto fail_arch;
    }
    break;

  default:
fail_arch:
    print_err("unsupported architecture (%s, [%u, %u])", bfd_info->printable_name, bfd_info->arch, bfd_info->mach);
    goto fail;
  }

  /* Symbol handling is best-effort only (they may not even be present) */
  load_symbols_bfd(bfd_h, bin);
  load_dynsym_bfd(bfd_h, bin);

  if(load_sections_bfd(bfd_h, bin) < 0) goto fail;

  /* Apply relocations if necessary */
  if (bin->arch == Binary::ARCH_PPC) {
    load_dynrelocs_bfd(bfd_h, bin);
  }

  ret = 0;
  goto cleanup;

fail:
  ret = -1;

cleanup:
  if(bfd_h) bfd_close(bfd_h);

  return ret;
}


int
load_binary_raw(std::string &fname, Binary *bin, Binary::BinaryType type)
{
  int ret;
  long fsize;
  FILE *f;
  Section *sec;

  f = NULL;

  bin->filename = std::string(fname);
  bin->type     = type;
  bin->type_str = std::string("raw");

  if(options.binary.arch == Binary::ARCH_NONE) {
    print_err("cannot determine binary architecture, specify manually");
    goto fail;
  }
  bin->arch     = options.binary.arch;
  bin->bits     = options.binary.bits;
  bin->arch_str = std::string(binary_arch_descr[(int)options.binary.arch][0]);
  bin->entry    = 0;

  if(!bin->bits) {
    switch(bin->arch) {
    case Binary::ARCH_X86:
      bin->bits = 64;
      break;
    default:
      break;
    }
  }

  bin->sections.push_back(Section());
  sec = &bin->sections.back();

  sec->binary = bin;
  sec->name   = std::string("raw");
  sec->type   = Section::SEC_TYPE_CODE;
  sec->vma    = options.binary.base_vma;

  f = fopen(fname.c_str(), "rb");
  if(!f) {
    print_err("failed to open binary '%s' (%s)", fname.c_str(), strerror(errno));
    goto fail;
  }

  fseek(f, 0L, SEEK_END);
  fsize = ftell(f);
  if(fsize <= 0) {
    print_err("binary '%s' appears to be empty", fname.c_str());
    goto fail;
  }

  sec->size  = (uint64_t)fsize;
  sec->bytes = (uint8_t*)malloc(fsize);
  if(!sec->bytes) {
    print_err("out of memory");
    goto fail;
  }

  fseek(f, 0L, SEEK_SET);
  if(fread(sec->bytes, 1, fsize, f) != (size_t)fsize) {
    print_err("failed to read binary '%s'", fname.c_str());
    goto fail;
  }

  ret = 0;
  goto cleanup;

fail:
  ret = -1;

cleanup:
  if(f) {
    fclose(f);
  }

  return ret;
}


int
load_binary(std::string &fname, Binary *bin, Binary::BinaryType type)
{
  if(type == Binary::BIN_TYPE_RAW) {
    return load_binary_raw(fname, bin, type);
  } else {
    return load_binary_bfd(fname, bin, type);
  }
}


void
unload_binary(Binary *bin)
{
  size_t i;
  Section *sec;

  for(i = 0; i < bin->sections.size(); i++) {
    sec = &bin->sections[i];
    if(sec->bytes) {
      free(sec->bytes);
    }
  }
}
